triage/berghelroach/berghelroach_test.go

```diff
@@ -0,0 +1,342 @@
+	genericLevenshteinVerify(t, "abcdefghijklmnopqrstuvwxyz",
+		"abcefghijklMnopqrStuvwxyz..",
+		5) // dMS..
+}
+
+// TestShortPattern tests Levenshtein edit distance on a very short pattern
+func TestShortPattern(t *testing.T) {
+	genericLevenshteinVerify(t, "short", "shirt", 1)
+}
+
+// TestZeroLengthPattern verifies zero-length behavior
+func TestZeroLengthPattern(t *testing.T) {
+	nonEmpty := "target"
+	genericLevenshteinVerify(t, "", nonEmpty, len(nonEmpty))
+	genericLevenshteinVerify(t, nonEmpty, "", len(nonEmpty))
+}
+
+// Utility functions
+
+// dynamicProgrammingLevenshtein computes Levenshtein edit distance
+// using the far-from-optimal dynamic programming technique.  This is
+// here purely to verify the results of better algorithms.
+func dynamicProgrammingLevenshtein(s1 string, s2 string) int {
+	lastRowSize := len(s1) + 1
+	lastRow := make([]int, lastRowSize)
+	for i := 0; i < lastRowSize; i++ {
+		lastRow[i] = i
+	}
+
+	for j := 0; j < len(s2); j++ {
+		thisRow := make([]int, len(lastRow))
+		thisRow[0] = j + 1
+		for i := 1; i < len(thisRow); i++ {
+			thisRow[i] = utils.Min(lastRow[i]+1,
+				thisRow[i-1]+1,
+				lastRow[i-1]+utils.BtoI(s2[j] != s1[i-1]))
+		}
+		lastRow = thisRow
+	}
+	return lastRow[len(lastRow)-1]
+}
+
+/*
+performEdits performs some edits on a string.
+
+original: string to be modified
+
+alphabet: some characters guaranteed not to be in the original
+
+replacements: how many single-character replacements to try
+
+insertions: how many characters to insert
+*/
+func performEdits(original string, alphabet string, replacements int, insertions int) (int, string) {
+	rand.Seed(768614336404564651)
+	edits := 0
+
+	// Convert the original string to a slice for easier manipulation.
+	originalSlice := []byte(original)
+
+	for i := 0; i < insertions; i++ {
+		utils.ByteSliceInsert(&originalSlice, alphabet[rand.Intn(len(alphabet))], rand.Intn(len(originalSlice)))
+		edits++
+	}
+
+	for i := 0; i < replacements; i++ {
+		where := rand.Intn(len(originalSlice))
+		letterInAlphabet := false
+		for i := 0; i < len(alphabet); i++ {
+			if originalSlice[where] == alphabet[i] {
+				letterInAlphabet = true
+				break
+			}
+		}
+
+		if !letterInAlphabet {
+			originalSlice[where] = alphabet[rand.Intn(len(alphabet))]
+			edits++
+		}
+	}
+
+	return edits, string(originalSlice)
+}
+
+/*
+generateRandomString Generates a long random alphabetic string,
+suitable for use with verifySomeEdits (using digits for the alphabet).
+
+size: desired string length
+
+seed: random number generator seed
+*/
+func generateRandomString(size int, seed int64) string {
+	alphabet := "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+
+	// Create a (repeatable) random string from the alphabet
+	rand.Seed(seed)
+	var builder strings.Builder
+
+	for i := 0; i < size; i++ {
+		builder.WriteByte(alphabet[rand.Intn(len(alphabet))])
+	}
+
+	return builder.String()
+}
+
+func getInstance(s string) berghelRoach {
+	return berghelRoach{pattern: s}
+}
```

Comment:
I honestly wasn't sure what to do about this one. The odd thing is that in Python, it would also create a new instance. I figured there was some reason it was done this way that I wasn't aware of. I can fix it up though/

Reply:
I'm guessing the python test may have been used to test other instances of edit distance algos at some point, and then reduced down to just this one?

Reply:
+1 https://github.com/kubernetes/test-infra/pull/18193/files#r453023869

Reply:
Addressed in #18881